core/pta/tests/aes_perf.c

```diff
@@ -0,0 +1,214 @@
+
+	/* Init ctx */
+	switch (algo) {
+	case TEE_ALG_AES_CBC_NOPAD:
+	case TEE_ALG_AES_CTR:
+	case TEE_ALG_AES_XTS:
+		iv = aes_iv;
+		iv_len = sizeof(aes_iv);
+		/*FALLTHROUGH*/
+	case TEE_ALG_AES_ECB_NOPAD:
+		res = crypto_cipher_init(*ctx, mode, aes_key, key_len, key2,
+					 key2_len, iv, iv_len);
+		break;
+	case TEE_ALG_AES_GCM:
+		res = crypto_authenc_init(*ctx, mode, aes_key, key_len, aes_iv,
+					  sizeof(aes_iv), TEE_AES_BLOCK_SIZE,
+					  0, payload_len);
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (res)
+		free_ctx(ctx, algo);
+
+	return res;
+}
+
+static TEE_Result update_ae(void *ctx, TEE_OperationMode mode,
+			    const void *src, size_t len, void *dst)
+{
+	size_t dlen = len;
+
+	return crypto_authenc_update_payload(ctx, mode, src, len, dst, &dlen);
+}
+
+static TEE_Result update_cipher(void *ctx, TEE_OperationMode mode,
+				const void *src, size_t len, void *dst)
+{
+	return crypto_cipher_update(ctx, mode, false, src, len, dst);
+}
+
+static TEE_Result do_update(void *ctx, uint32_t algo, TEE_OperationMode mode,
+			    unsigned int rep_count, unsigned int unit_size,
+			    const uint8_t *in, size_t sz, uint8_t *out)
+{
+	TEE_Result (*update_func)(void *ctx, TEE_OperationMode mode,
+				  const void *src, size_t len,
+				  void *dst) = NULL;
+	TEE_Result res = TEE_SUCCESS;
+	unsigned int n = 0;
+	unsigned int m = 0;
+
+	if (algo == TEE_ALG_AES_GCM)
+		update_func = update_ae;
+	else
+		update_func = update_cipher;
+
+	for (n = 0; n < rep_count; n++) {
+		for (m = 0; m < sz / unit_size; m++) {
+			res = update_func(ctx, mode, in + m * unit_size,
+					  unit_size, out + m * unit_size);
+			if (res)
+				return res;
+		}
+		if (sz % unit_size)
+			res = update_func(ctx, mode, in + m * unit_size,
+					  sz % unit_size, out + m * unit_size);
+	}
+
+	return res;
+}
+
+TEE_Result core_aes_perf_tests(uint32_t param_types,
+			       TEE_Param params[TEE_NUM_PARAMS])
+{
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INOUT,
+						   TEE_PARAM_TYPE_MEMREF_INOUT);
+	TEE_Result res = TEE_SUCCESS;
+	TEE_OperationMode mode = 0;
+	unsigned int rep_count = 0;
+	unsigned int unit_size = 0;
+	size_t key_size_bits = 0;
+	uint32_t algo = 0;
+	void *ctx = NULL;
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	switch (params[0].value.b) {
+	case PTA_INVOKE_TESTS_MODE_ECB:
+		algo = TEE_ALG_AES_ECB_NOPAD;
+		break;
+	case PTA_INVOKE_TESTS_MODE_CBC:
+		algo = TEE_ALG_AES_CBC_NOPAD;
+		break;
+	case PTA_INVOKE_TESTS_MODE_CTR:
+		algo = TEE_ALG_AES_CTR;
+		break;
+	case PTA_INVOKE_TESTS_MODE_XTS:
+		algo = TEE_ALG_AES_XTS;
+		break;
+	case PTA_INVOKE_TESTS_MODE_GCM:
+		algo = TEE_ALG_AES_GCM;
+		break;
+	default:
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (params[0].value.a >> 16)
+		mode = TEE_MODE_DECRYPT;
+	else
+		mode = TEE_MODE_ENCRYPT;
+
+	key_size_bits = params[0].value.a & 0xffff;
```

Comment:
`GENMASK_32(15, 0)`

Reply:
For odd shaped masks I would agree, but this is like `GENMASK_32(7, 0)` instead of `0xff`. Anyone doing bitmasking must surely instantly recognize a 16-mask if written as `0xffff`.